Fix mouse button behavior in Master of Orion II

FeralChild64 opened the pull request:
Fixes mouse button behavior in Master of Orion II, reported by @brandonrobertz.

Reason: the way game interacts with DOS mouse driver (it changes the event callback settings by calling INT 33 function 0x0F constantly) confused the recently introduced optimizations, which skipped pushing events through internal queue when it was thought to be safe. Optimizations had to be removed.

Grounded0 commented:
The mouse works as it should now with this fix.

johnnovak commented on src/ints/mouse.cpp:
```diff
@@ -949,8 +949,8 @@ void MOUSE_EventReleased(uint8_t idx)
-        MOUSESERIAL_NotifyPressedReleased(buttons_12S, idx_12S);
+	event.request_dos = MOUSEDOS_NotifyReleased(buttons_12S, idx_12S);
+	MOUSESERIAL_NotifyPressedReleased(buttons_12S, idx_12S);
     }
```
Indentation seems to be off in this file.

FeralChild64 replied:
Fixed.

johnnovak commented on src/ints/mouse_dos_driver.cpp:
```diff
@@ -866,19 +866,18 @@ bool MOUSEDOS_NotifyMoved(const float x_rel, const float y_rel,
+    // is at least one game (Master of Orion II), which does INT 0x33
+    // calls with 0x0F parameter (changing the callback settings)
+    // constantly (don't ask me, why) - doing too much optimization
+    // can cause the game to skip mouse events.
```
Nice comment for the next person who's going to work in this area 🥇

johnnovak reviewed (approved):
Looking good 👍🏻

kcgen commented:
This addresses the buttons; I checked if movement events are also affected, but it looks like it always returns true on positive movement regardless of the callback state (even if the callback mask is uninterested in the event):

``` c++
// If we are here, there is some noticeable change in mouse
// state - if callback is registered for mouse movement,
// then we definitely need the event
if (MOUSEDOS_HasCallback(static_cast<uint8_t>(MouseEventId::MouseHasMoved)))
    return true;

// Noticeable change, but no callback; we might still need the
// event for cursor redraw routine - check this
return abs_changed && !state.hidden && !state.inhibit_draw;
```

That last comment is interesting: _"Noticeable change, but no callback; we might still need the event for cursor redraw routine"_

I suspect, what this issue/PR has told us, is that the same situation applies to buttons? That is, _we might still need the (button) event for cursor redraw routines_.

kcgen commented:
@FeralChild64 , I noticed the callback mask type is uint16_t, however we only ever call it with 8-bits (and our enum also only holds 8 events).

However, when we step up to MS MOUSE 3.0+, it's callback event mask register is 16-bits, however our `callback_mask & event_id_mask` operation will always return false if it's watching for event-bits 9 through 15. 

Can you tell if MOO2 is maybe watching for any of these upper bit events? 

Did MS MOUSE 3.0 generate any of those upper-bit events for basic button and movement events?

kcgen reviewed (approved):
Thanks for the super quick update, @FeralChild64.

Just a couple unrelated question (while we've got eyes in this area!)

Once the formatting is fixed up, feel free to merge.

FeralChild64 commented:
@kcgen There is no need to store more than 8 bits of the mask. Usually just 5 bits are needed (movement, left button pressed, left button released, right button pressed, right button released). Probably most 3-button mouse drivers use 2 more bits: middle button pressed, middle button released. CuteMouse WheelApi adds one more bit, for wheel movements - that's 8 bits total.

See here:
- https://stanislavs.org/helppc/int_33-c.html
- https://stanislavs.org/helppc/int_33-14.html
- https://github.com/FDOS/mouse/blob/master/int33.lst
- https://github.com/FDOS/mouse/blob/master/wheelapi.txt

I don't expect any existing software to listen to bits above the 8 mentioned.

The comment _Noticeable change, but no callback; we might still need the event for cursor redraw routine_ is now removed, with the whole checks. Right now, there is only one optimization left:
- if the mouse movement is 'sub-pixel' - no need to issue any callback, guest software won't notice anything
But previously I was checking whether we have a cursor drawn by the mouse driver (in such case we need INT74, as the cursor position is updated there) or if there was a callback registered for the mouse movement.

MOO2 does something I totally don't understand - it constantly changes the events it is interested in, this + my optimizations caused it to miss the mouse release events, and possibly some mouse movements too. Therefore, I removed the optimizations. It's possible I could have removed less - but there might be more games doing similar things out there, I prefer to have a safer code.

kcgen commented:
Thanks for this thorough explanation @FeralChild64!

> it constantly changes the events it is interested in

Maybe it's a coding bug, and believes the events will be "boolean-or" on the DOS side? Otherwise I agree.. that really doesn't make sense, given it cannot predict what the user will do w/ the mouse..

brandonrobertz commented:
Amazing work @FeralChild64 👏
